client-api,events: Remove `StrippedState` and add optional fields to `StrippedStateEvent` instead

zecakeh opened the pull request:
`StrippedState` makes it complicated to access the content of an event, with several levels of matching (event format, event type, and for some formats another level), to end up with up to 3 possible content types.

It is simpler for users to have a single enum for the event type and be able to access the content and the fields directly.

This requires a first commit to add support for optional fields to the `Event` derive macro.

zecakeh committed:
events: Add `default` attribute for Event derive macro

Allows to use the `Default` implementation of a field if it is missing.

Kladki commented on crates/ruma-events/CHANGELOG.md:
```diff
@@ -98,6 +98,10 @@ Improvements:
 - Add `additional_creators` field to `RoomCreateEventContent`, used to optionally specify
   additional creators of a room.
+- Add unstable support for events with the same format as `StateEvent` in `StrippedStateEvent`,
+  according to MSC4311.
```
MSC4311 now just uses the classic stripped state format.

zecakeh replied:
Okay, I changed the commit history to remove it first in `StrippedState`, and then remove `StrippedState` in favor of extra fields in `StrippedStateEvent`.

zecakeh committed:
client-api: Remove support for previous version of MSC4311

The MSC was updated and the `m.room.create` event now has the same
format as other stripped state events in the CS API.

zecakeh committed:
client-api,events: Remove StrippedState and add optional fields to StrippedStateEvent instead

`StrippedState` makes it complicated to access the content of an event,
with several levels of matching (event format, event type, and for some
formats another level), to end up with up to 3 possible content types.

It is simpler for users to have a single enum for the event type and be
able to access the content and the fields directly.

Kladki commented on crates/ruma-events/src/kinds.rs:
```diff
@@ -446,6 +446,20 @@ pub struct StrippedStateEvent<C: PossiblyRedactedStateEventContent> {
     /// affects.
     pub state_key: C::StateKey,
+
+    /// The globally unique event identifier for the user who sent the event.
```
> for the user who sent the event

This doesn't seem right, shouldn't it just be
```suggestion
    /// The globally unique identifier for the event.
```

zecakeh replied:
Oops, indeed.

zecakeh replied:
That said, I copy-pasted it from another definition :stuck_out_tongue:.

I'm going to fix them in another commit.

zecakeh replied:
Actually, that makes me want to fix other docs, I'll do it in a separate PR.
